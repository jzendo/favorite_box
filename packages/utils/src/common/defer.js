export default () => {
  let resolve
  let reject

  // eslint-disable-next-line
  const promise = new Promise((resolve_, reject_) => {
    resolve = resolve_
    reject = reject_
  })

  return {
    promise,
    resolve,
    reject
  }
}
