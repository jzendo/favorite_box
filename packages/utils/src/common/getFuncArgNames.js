import getFuncParameterNames from 'get-parameter-names'

export default getFuncParameterNames
