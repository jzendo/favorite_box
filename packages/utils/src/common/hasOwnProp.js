const objectProtoHasOwn = Object.prototype.hasOwnProperty

export default (obj, p) => {
  if (!obj || p === undefined) return false

  return objectProtoHasOwn.call(obj, p)
}
