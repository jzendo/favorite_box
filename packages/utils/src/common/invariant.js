// Influenced by https://github.com/facebook/react/blob/v16.13.0/packages/shared/invariant.js

export default function invariant (condition, format, a, b, c, d, e, f) {
  if (process.env.NODE_ENV === 'production') return

  if (arguments.length < 2) {
    throw new Error('invalid argument count.')
  }

  if (!condition) {
    const args = [a, b, c, d, e, f]
    let argIndex = 0
    const error = new Error(format.replace(/%s/g, () => args[argIndex++]))
    error.name = 'Invariant Violation'
    throw error
  }
}
