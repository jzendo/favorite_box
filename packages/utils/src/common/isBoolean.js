import typeIs from './typeIs'

export default obj => typeIs(obj, 'boolean')
