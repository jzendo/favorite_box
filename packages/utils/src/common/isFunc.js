export default test => typeof test === 'function'
