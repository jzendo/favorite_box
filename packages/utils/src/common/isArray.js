import isFunc from './isFunc'

export default test => {
  if (isFunc(Array.isArray)) return Array.isArray(test)
  return typeof test === 'object' && test.constructor === Array
}
