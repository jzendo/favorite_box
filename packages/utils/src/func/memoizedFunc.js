import getFuncArgNames from '../common/getFuncArgNames'
import hasOwnProperty from '../common/hasOwnProp'
import isFunc from '../common/isFunc'
import stringify from '../common/stringify'

const getInitCache = () => Object.create(null)

const generateCacheUtil = () => {
  let cached = getInitCache()

  return {
    has (key) {
      if (!key || cached[key] === undefined) return false

      return hasOwnProperty(cached, key)
    },
    get (key) {
      if (this.has(key)) {
        return cached[key]
      }

      return undefined
    },
    add (key, val) {
      cached[key] = val
      return this
    },
    del (key) {
      if (this.has(key)) {
        cached[key] = undefined
      }
      return this
    },
    clear () {
      cached = getInitCache()
      return this
    }
  }
}

const genDyncFnWithOriginArgNames = (argNames, fnName) => {
  const helpArgsNames = ['fn', 'stringify', 'cacheUtil']

  const wrapFnName = `memoized${fnName.slice(0, 1).toUpperCase() +
    fnName.slice(1)}`

  const strSameAsCondition = argNames
    .map(arg => `prevArgs.${arg} === ${arg}`)
    .join(' && ')

  const strStoreArgCondition = argNames
    .map(arg => `;prevArgs.${arg} = ${arg}`)
    .join('')

  const strMap =
    '{' +
    argNames.join(': undefined, ') +
    (argNames.length ? ': undefined' : '') +
    '}'

  // eslint-disable-next-line
  return new Function(
    ...helpArgsNames,
    `
/*!
 * The function which is memoized is named '${fnName}'
 */

var disposed = false

var prev_uniqueArgKey
var prevArgs = ${strMap}

var wrapFn = function ${wrapFnName}(${argNames.join(', ')}) {
  var uniqueArgKey

  if (!disposed && prevArgs) {
    if (${strSameAsCondition}) {
      uniqueArgKey = prev_uniqueArgKey
    } else {
      uniqueArgKey = stringify(${argNames.join(', ')})
      prev_uniqueArgKey = uniqueArgKey
      ${strStoreArgCondition}
    }
  }

  if (!disposed && cacheUtil.has(uniqueArgKey)) {
    ${
      /* Only for unit/test, do not modify here!! */
      process.env.NODE_ENV === 'test'
        ? `fn.__onCacheGetter([${argNames.join(
            ', '
          )}], uniqueArgKey, cacheUtil)`
        : ''
    }

    return cacheUtil.get(uniqueArgKey)
  }${
    process.env.NODE_ENV === 'test'
      ? `else { fn.__onSkipGetter(disposed, [${argNames.join(
          ', '
        )}], uniqueArgKey, cacheUtil) }`
      : ''
  }

  var args = [${argNames.join(', ')}]
  var r = fn(${argNames.join(', ')})

  if (!disposed && r !== undefined) {
    cacheUtil.add(uniqueArgKey, r)
    ${
      process.env.NODE_ENV === 'test'
        ? `fn.__onCacheSetter(r, [${argNames.join(
            ', '
          )}], uniqueArgKey, cacheUtil)`
        : ''
    }
  }${
    process.env.NODE_ENV === 'test'
      ? `else { fn.__onSkipSetter(disposed, r, [${argNames.join(
          ', '
        )}], uniqueArgKey, cacheUtil) }`
      : ''
  }

  return r
}

wrapFn.dispose = () => {
  disposed = true
  prevArgs = null
}

return wrapFn
  `
  )
}

const keyOfMemoizedFn =
  typeof Symbol !== 'undefined'
    ? Symbol('Symbol.memoizedFunc')
    : 'Symbol.memoizedFunc'

/**
 * Get function with memoized the function's result for performance.
 *
 * @param {function} fn the function will be memoized.
 * @returns {function}
 */
export default (fn, userStringify) => {
  if (!isFunc(fn)) {
    throw new TypeError('The first parameter should be a function.')
  }
  if (fn.length === 0) {
    throw new TypeError('The function should be one parameter at least.')
  }

  let memoizedFn = fn[keyOfMemoizedFn]
  // Function name
  const fnName = fn.name || '<anonymous>'

  if (!memoizedFn) {
    const argNames = getFuncArgNames(fn)
    memoizedFn = genDyncFnWithOriginArgNames(argNames, fnName)
    fn[keyOfMemoizedFn] = memoizedFn

    if (
      process.env.NODE_ENV !== 'production' &&
      process.env.NODE_ENV !== 'test'
    ) {
      console.log(`The memoized function: ${fnName}`)
    }
  }

  // Use user `stringify` as possible as can
  // But the default is implemented by `JSON.stringify` that is not high efficiency
  // TODO: use better `stringify` implement
  const stringifyHandler = (...args) => {
    if (userStringify) return userStringify(...args)
    return stringify(args)
  }

  return memoizedFn(fn, stringifyHandler, generateCacheUtil())
}
