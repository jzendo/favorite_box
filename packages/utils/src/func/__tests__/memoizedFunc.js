/* global describe, test, expect */
import memoizedFunc from '../memoizedFunc'
// import getFuncArgNames from '../../common/getFuncArgNames'
import { applyTestFnForSingleParam } from './common/memoizedFuncHelper'

describe('memoizedFunc', () => {
  describe('single param func', () => {
    const singleParamArrowFn = a => a
    applyTestFnForSingleParam(singleParamArrowFn)

    // eslint-disable-next-line
    const singleParamFn = function singleParamFn (/* a) */ a) {
      return a
    }
    applyTestFnForSingleParam(singleParamFn)
  })

  describe('multi params func', () => {
    // eslint-disable-next-line
    const singleParamArrowFn = (a, b) => a
    applyTestFnForSingleParam(singleParamArrowFn)

    // Eslint should be disable for next line, for `user dirty function declare`
    // eslint-disable-next-line
    const singleParamFn = function singleParamFn (/* a) */ a, b) {
      return a
    }
    applyTestFnForSingleParam(singleParamFn)
  })

  test('the first parameter should be function', () => {
    expect(() => {
      memoizedFunc()
    }).toThrow()

    expect(() => {
      memoizedFunc(1)
    }).toThrow()

    expect(() => {
      memoizedFunc(true)
    }).toThrow()
  })
})
