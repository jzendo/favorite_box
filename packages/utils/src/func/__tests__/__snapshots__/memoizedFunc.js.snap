// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`memoizedFunc multi params func func name: singleParamArrowFn 1`] = `
Array [
  "a",
  "b",
]
`;

exports[`memoizedFunc multi params func func name: singleParamFn 1`] = `
Array [
  "a",
  "b",
]
`;

exports[`memoizedFunc single param func func name: singleParamArrowFn 1`] = `
Array [
  "a",
]
`;

exports[`memoizedFunc single param func func name: singleParamFn 1`] = `
Array [
  "a",
]
`;
